DIPs/DIP1001.md

```diff
@@ -0,0 +1,291 @@
+# Exception Handling Extensions
+
+| Section         | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            | 1001                                                            |
+| Author:         | Freddy A Cubas (hexagonalstar64@gmail.com)                      |
+| Implementation: | (Nothing yet)                                                   |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+* Add new @throws(types...) attribute
+* Allows functions to throw any type as a exception
+* Add a way to specify whether generated error handling code should use D style
+stack unwinding or C style error codes or a mix of both
+
+### Links
+
+* [Current property syntax](https://dlang.org/spec/attribute.html#PropertyIdentifier)
+* [Current function mangling](https://dlang.org/spec/abi.html#FuncAttr)
+
+## Description
+
+#### Throws Property
+	PropertyIdentifier:
+		property
+		safe
+		trusted
+		system
+		disable
+		nogc
+		throws()
+		throws(TypeList)
+		throws(auto)
+	TypeList:
+		Type
+		Type ,
+		Type , TypeList
+
+The throws attribute specifies that a function may throw any of the types
+specified in TypeList.
+
+@throws() is equivalent to nothrow.
+
+By default all functions are @throws(Throwable).
+
+@throws(auto) infers all thrown types
+
+functions that return auto or function templates and don't have a explict
+@throws() have @throws(auto).
+
+A function which calls a sub function with a @throws(TypeList) attribute must
+declare all uncaught exceptions within its own @throws(TypeList) attribute of
+the caller function.
+
+##### Mangling
+
+Mangling:
+
+	FuncAttr:
+		empty
+		FuncAttrPure
+		FuncAttrNogc
+		FuncAttrNothrow
+		FuncAttrProperty
+		FuncAttrRef
+		FuncAttrReturn
+		FuncAttrScope
+		FuncAttrTrusted
+		FuncAttrSafe
+		FuncAttrThrows
+	FuncAttrThrows:
+		Nt Number ThrownTypes
+	ThrownTypes:
+		Type
+		Type ThrownTypes
+
+Number is the number of elements in a @throws(TypeList) attributes, ThrownTypes
+are the list of types in TypeList.
+
+If a @throws() attribute is empty, FuncAttrNothrow should be generated instead.
+If a @throws() attribute just contains Throwable, empty should be generated
+instead.
+
+
+#### Throw Any Type
+
+Any type may be thrown not just classes that inherit from Throwable.
+
+A thrown type must be implictly convertable to one of the types mentioned in
+the @throws(TypeList) list.
+
+This fits in with the old class based exception handling.
+
+A possible optimization would be to generate a seperate unwinding table for
+each value type.
+
+#### Error Code Unwinding
+
+For all Types in a @throws(TypeList) attribute that have a manifest
+constant(__ErrorCode) and that constant evaluates to true, the compiler should
+generate C style error code handling for that function using a tagged union of
+the result and all __ErrorCode exceptions.
```

Comment:
This can already be done with current D capability as a library. Why does it need to be backed in the language ?